### PR TITLE
ISPN-8336 lockAllAndRecord() method is blocking

### PR DESCRIPTION
**Pull request by rvansa**
Rebased #5515 and hacked hibernate-cache

### CODE REVIEW

**Comment by rvansa** [comment]
@pruivo Check this out... I've accidentally mixed my changes into your commit, could separate it later on.

File: hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/NonTxInvalidationCacheAccessDelegate.java [hunk 1]
```diff
@@ -7,17 +7,36 @@
...
  * @author Radim Vansa &lt;rvansa@redhat.com&gt;
  */
 public class NonTxInvalidationCacheAccessDelegate extends InvalidationCacheAccessDelegate {
+	private final AsyncInterceptorChain invoker;
+	private final CommandsFactory commandsFactory;
+	private final Metadata metadata;
+
 	public NonTxInvalidationCacheAccessDelegate(BaseRegion region, PutFromLoadValidator validator) {
 		super(region, validator);
+		ComponentRegistry cr = region.getCache().getComponentRegistry();
```
Comment (pruivo) [inline hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/NonTxInvalidationCacheAccessDelegate.java]: this was the kind of code I didn't want to see :(

I'm not sure if there are other users who are using ThreadLocal and pointing them this code if their code stops working doesn't seem right

I guess we don't have a choice right?

Comment (rvansa) [reply hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/NonTxInvalidationCacheAccessDelegate.java]: 2LC implementation does worse things than this. I haven't changed any Infinispan API. And it seems to me that this is better than actually tunneling any information between different components through `ThreadLocal`.

There can't be other consumers of the ThreadLocal - note that if anyone started to hack 2LC even more, the PFLV has changed package so it's broken anyway. IMO the 2LC is not meant as anything 'extensible' (beyond e.g. wfly integration), it's really an implementation of component, not a library.

**Comment by rvansa** [comment]
run performance tests please
library

**Comment by rvansa** [comment]
@danberindei I've handled some outstanding comments in #5515 so if you find this ok and both CI and PerfAck give it green, please merge it (I nor Pedro can't since we're both authors).

**Comment by rvansa** [comment]
run performance tests please
library

**Comment by rvansa** [comment]
@danberindei The perf ack does not show any substantial difference.

**Comment by rvansa** [comment]
@pruivo Hi, since it's too hard to get Dan's attention, please check my changes and if you're ok with these, I'll rebase and let CI run once more and you can merge that yourselves.

**Comment by rvansa** [comment]
I've had some failures in `ProtobufRemoteIteratorIndexingTest` related to the fact that the test uses ForkJoin common pool to execute blocking calls. As the `thenApplyAsync` in locking uses common pool, too, this lead to deadlock. Therefore I need to update this all to use remote executor instead.

**Comment by pruivo** [comment]
@rvansa ok! let me know when it is updated.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java [hunk 2]
```diff
@@ -56,6 +62,8 @@
    @Inject private Configuration configuration;
    @Inject @ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR)
    private ScheduledExecutorService scheduler;
+   @Inject @ComponentName(KnownComponentNames.REMOTE_COMMAND_EXECUTOR)
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java]: :-1: 
use the `ASYNC_OPERATIONS_EXECUTOR`. 
The `REMOTE_COMMAND_EXECUTOR` will probably be removed in the future.

File: hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/InvalidationCacheAccessDelegate.java [hunk 3]
```diff
@@ -123,16 +123,10 @@ public boolean putFromLoad(Object session, Object key, Object value, long txTime
...
-		try {
-			// We update whether or not the region is valid. Other nodes
-			// may have already restored the region so they need to
-			// be informed of the change.
-			writeCache.remove(key);
-		}
-		finally {
-			putValidator.resetCurrentSession();
-		}
+		// We update whether or not the region is valid. Other nodes
```
Comment (pruivo) [inline hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/InvalidationCacheAccessDelegate.java]: aren't you missing the `session`?

File: hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/LockingInterceptor.java [hunk 4]
```diff
@@ -41,11 +41,22 @@
          throw throwable;
 
       if (rv instanceof CompletableFuture) {
+         // The future is produced by produced in UnorderedDistributionInterceptor,
```
Comment (pruivo) [inline hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/LockingInterceptor.java]: typo: produced by produced

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java [hunk 5]
```diff
@@ -20,14 +23,16 @@
...
+   private Executor executor;
    private TimeService timeService;
 
    public PerKeyLockContainer() {
       lockMap = new ConcurrentHashMap<>(INITIAL_CAPACITY);
    }
 
    @Inject
-   public void inject(TimeService timeService) {
+   public void inject(@ComponentName(KnownComponentNames.REMOTE_COMMAND_EXECUTOR) Executor executor, TimeService timeService) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java]: :-1: for `REMOTE_COMMAND_EXECUTOR`

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java [hunk 6]
```diff
@@ -28,10 +31,10 @@ public StripedLockContainer(int concurrencyLevel) {
    }
 
    @Inject
-   public void inject(TimeService timeService) {
+   public void inject(@ComponentName(KnownComponentNames.REMOTE_COMMAND_EXECUTOR) Executor executor, TimeService timeService) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java]: :-1: for `REMOTE_COMMAND_EXECUTOR`

File: hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/SessionInvocationContext.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+      if (this.key != null && isKeyEquals(key)) {
+         this.cacheEntry = null;
+      }
+   }
+
+   public Object getKey() {
+      return key;
+   }
+
+   public CacheEntry getCacheEntry() {
```
Comment (pruivo) [inline hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/SessionInvocationContext.java]: unused

File: hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/SessionInvocationContext.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+   public boolean isEntryRemovedInContext(final Object key) {
+      CacheEntry ce = lookupEntry(key);
+      return ce != null && ce.isRemoved() && ce.isChanged();
+   }
+
+   public Object getSession() {
+      return session;
+   }
+
+   public void resetState() {
```
Comment (pruivo) [inline hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/SessionInvocationContext.java]: unused

File: hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/InvalidationCacheAccessDelegate.java (hunk 3 above)
Comment (rvansa) [reply hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/InvalidationCacheAccessDelegate.java]: No, this gets overridden in NonTxInvalidationCacheAccessDelegate. Transactional variant does not need the session.

**Comment by rvansa** [comment]
@pruivo Addressed.

**Comment by pruivo** [comment]
@rvansa there are failures in CI a bit suspicious. can you take a look? can it be related to the PR?

**Comment by rvansa** [comment]
@pruivo Do you mean the 2LC failure? That looks similar to https://github.com/infinispan/infinispan/pull/5673#issuecomment-358240582 (though here it's non-JTA). I'll trigger one more build.

**Comment by rvansa** [comment]
@pruivo Hi, the `IndexingDuringStateTransferTest.testMergeRemove` failure you've mentioned is caused by functional triangle, not locking: see http://ci.infinispan.org/job/Infinispan/job/PR-5703/7/ - it's been there, and was merged.

**Pull request merged**